Source/Core/DolphinLibretro/Input.cpp

```diff
@@ -655,48 +717,81 @@ void retro_set_controller_port_device(unsigned port, unsigned device)
     switch (device)
     {
     case RETRO_DEVICE_WIIMOTE:
-      desc = Libretro::Input::descWiimote;
       wmExtension->SetSelectedAttachment(ExtensionNumber::NONE);
       WiimoteCommon::SetSource(port, WiimoteSource::Emulated);
       break;
 
     case RETRO_DEVICE_WIIMOTE_SW:
-      desc = Libretro::Input::descWiimoteSideways;
       wmExtension->SetSelectedAttachment(ExtensionNumber::NONE);
       static_cast<ControllerEmu::NumericSetting<bool>*>(wmOptions->numeric_settings[2].get())
           ->SetValue(true);  // Sideways Wiimote
       WiimoteCommon::SetSource(port, WiimoteSource::Emulated);
       break;
 
     case RETRO_DEVICE_WIIMOTE_NC:
-      desc = Libretro::Input::descWiimoteNunchuk;
       wmExtension->SetSelectedAttachment(ExtensionNumber::NUNCHUK);
       WiimoteCommon::SetSource(port, WiimoteSource::Emulated);
       break;
 
     case RETRO_DEVICE_WIIMOTE_CC:
     case RETRO_DEVICE_WIIMOTE_CC_PRO:
-      desc = Libretro::Input::descWiimoteCC;
       wmExtension->SetSelectedAttachment(ExtensionNumber::CLASSIC);
       WiimoteCommon::SetSource(port, WiimoteSource::Emulated);
       break;
 
     case RETRO_DEVICE_REAL_WIIMOTE:
-      desc = Libretro::Input::descEmpty;
       WiimoteCommon::SetSource(port, WiimoteSource::Real);
 
     default:
-      desc = Libretro::Input::descGC;
       WiimoteCommon::SetSource(port, WiimoteSource::None);
       break;
     }
     wm->UpdateReferences(g_controller_interface);
     ::Wiimote::GetConfig()->SaveConfig();
   }
+  std::vector<retro_input_descriptor> all_descs;
 
-  retro_input_descriptor* desc_ptr = desc;
-  while (desc_ptr->description)
-    (desc_ptr++)->port = port;
+  int port_max = SConfig::GetInstance().bWii ? 8 : 4;
+  for (int i = 0; i < port_max; i++)
+  {
+    retro_input_descriptor* desc;
 
-  Libretro::environ_cb(RETRO_ENVIRONMENT_SET_INPUT_DESCRIPTORS, desc);
+    switch (Libretro::Input::input_types[i])
+    {
+    case RETRO_DEVICE_WIIMOTE_SW:
+      desc = Libretro::Input::descWiimoteSideways;
+      break;
+
+    case RETRO_DEVICE_WIIMOTE_NC:
+      desc = Libretro::Input::descWiimoteNunchuk;
+      break;
+
+    case RETRO_DEVICE_WIIMOTE_CC:
+    case RETRO_DEVICE_WIIMOTE_CC_PRO:
+      desc = Libretro::Input::descWiimoteCC;
+      break;
+
+    case RETRO_DEVICE_NONE:
+      continue;
+
+    default:
+      if (!SConfig::GetInstance().bWii || port > 3)
+      {
+        desc = Libretro::Input::descGC;
+      }
+      else
+      {
+        desc = Libretro::Input::descWiimote;
+      }
+      break;
+    }
+    for (int j = 0; desc[j].device != 0; j++)
+    {
+      retro_input_descriptor new_desc = desc[j];
+      new_desc.port = i;
+      all_descs.push_back(new_desc);
+    }
+  }
+  all_descs.push_back({0});
+  Libretro::environ_cb(RETRO_ENVIRONMENT_SET_INPUT_DESCRIPTORS, &all_descs[0]);
 }
```

Comment:
I think you forgot `RETRO_DEVICE_REAL_WIIMOTE` here.
you probably don't event need to duplicate the logic with an new switch block, it would actually make more sense to just keep the `all_descs` array static and update only one port at a time.